lib/trino-parquet/src/main/java/io/trino/parquet/reader/ChunkedInputStream.java

```diff
@@ -0,0 +1,119 @@
+import static com.google.common.base.Preconditions.checkPositionIndexes;
+import static com.google.common.io.ByteStreams.readFully;
+import static java.util.Objects.requireNonNull;
+
+public class ChunkedInputStream
+        extends InputStream
+{
+    private final LocalMemoryContext dataReaderMemoryUsage;
+    private final Iterator<BasicSliceInput> chunks;
+    private BasicSliceInput current;
+
+    public ChunkedInputStream(Collection<ChunkReader> chunks, AggregatedMemoryContext memoryContext)
+    {
+        this.dataReaderMemoryUsage = requireNonNull(memoryContext, "memoryContext is null")
+                .newLocalMemoryContext(ChunkedInputStream.class.getSimpleName());
+        requireNonNull(chunks, "chunks is null");
+        checkArgument(!chunks.isEmpty(), "At least one chunk is expected but got none");
+        this.chunks = Iterators.transform(chunks.iterator(), this::readChunk);
+        current = this.chunks.next();
+    }
+
+    public Slice getSlice(int length)
+    {
+        if (length == 0) {
+            return Slices.EMPTY_SLICE;
+        }
+        while (!current.isReadable()) {
+            checkArgument(chunks.hasNext(), "Requested %s bytes but 0 was available", length);
+            current = chunks.next();
+        }
+        if (current.available() >= length) {
+            return current.readSlice(length);
+        }
+        // requested length crosses the slice boundary
+        byte[] bytes = new byte[length];
+        try {
+            readFully(this, bytes, 0, bytes.length);
+        }
+        catch (IOException e) {
+            throw new RuntimeException("Failed not read " + length + " bytes", e);
+        }
+        return Slices.wrappedBuffer(bytes);
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len)
+    {
+        checkPositionIndexes(off, off + len, b.length);
+        if (len == 0) {
+            return 0;
+        }
+        while (!current.isReadable()) {
+            if (!chunks.hasNext()) {
+                return -1;
+            }
+            current = chunks.next();
+        }
+
+        return current.read(b, off, len);
+    }
+
+    @Override
+    public int read()
+            throws IOException
+    {
+        while (!current.isReadable() && chunks.hasNext()) {
+            current = chunks.next();
+        }
+
+        return current.read();
+    }
+
+    @Override
+    public int available()
+            throws IOException
+    {
+        return current.available();
+    }
+
+    // potentially we can also implement skip here that is more performant that the default implementation
+
+    private BasicSliceInput readChunk(ChunkReader chunkReader)
+    {
+        Slice slice = chunkReader.readUnchecked();
+        checkArgument(slice.length() > 0, "all chunks have to be not empty");
+        // todo this just an estimate and doesn't reflect actual retained memory
+        dataReaderMemoryUsage.setBytes(slice.length());
```

Comment:
I think this should be `slice.getRetainedSize` to account for the possibility that the underlying slice is holding on to more data due to merging nearby reads.